File: src/lib/wow/archive/casc/cdn-config.ts
/**
 * Parser for CDN config files (BuildConfig / CDNConfig),
 * ported from wow.export (src/js/casc/cdn-config.js).
 */
const KEY_VAR_PATTERN = /([^\s]+)\s?=\s?(.*)/;

/**
 * Convert a config key such as 'encoding-sizes' to 'encodingSizes'.
 */
function normalizeKey(key: string): string {
  const keyParts = key.split('-');

  if (keyParts.length === 1) return key;

  for (let i = 1, n = keyParts.length; i < n; i++) {
    const part = keyParts[i];
    keyParts[i] = part.charAt(0).toUpperCase() + part.slice(1);
  }

  return keyParts.join('');
}

export type CDNConfigEntries = Record<string, string>;

export function parseCDNConfig(data: string): CDNConfigEntries {
  const entries: CDNConfigEntries = {};
  const lines = data.split(/\r?\n/);

  // Valid config files start with "# " (e.g., "# Build Configuration", "# CDN Configuration")
  const hasValidHeader = lines.length > 0 && lines[0].trim().startsWith('# ');

  if (!hasValidHeader) throw new Error('Invalid CDN config: unexpected start of config');

  for (const line of lines) {
    // Skip empty lines/comments.
    if (line.trim().length === 0 || line.startsWith('#')) continue;

    const match = line.match(KEY_VAR_PATTERN);
    if (match === null) throw new Error('Invalid token encountered parsing CDN config');

    entries[normalizeKey(match[1])] = match[2];
  }

  return entries;
}

export default parseCDNConfig;

File: golang/internal/wow/archive/casc/cdn_config.go
package casc

import (
	"regexp"
	"strings"
)

// CDNConfigEntries maps normalized config keys to values.
type CDNConfigEntries map[string]string

var keyVarPattern = regexp.MustCompile(`([^\s]+)\s?=\s?(.*)`)

func normalizeCDNKey(key string) string {
	parts := strings.Split(key, "-")
	if len(parts) == 1 {
		return key
	}
	for i := 1; i < len(parts); i++ {
		if len(parts[i]) > 0 {
			parts[i] = strings.ToUpper(parts[i][:1]) + parts[i][1:]
		}
	}
	return strings.Join(parts, "")
}

// ParseCDNConfig parses a CDN config file.
func ParseCDNConfig(data string) (CDNConfigEntries, error) {
	entries := CDNConfigEntries{}
	lines := strings.Split(data, "\n")
	hasValidHeader := len(lines) > 0 && strings.HasPrefix(strings.TrimSpace(lines[0]), "# ")
	if !hasValidHeader {
		return nil, errInvalidCDNConfig
	}
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(line, "#") {
			continue
		}
		match := keyVarPattern.FindStringSubmatch(line)
		if match == nil {
			return nil, errInvalidCDNToken
		}
		entries[normalizeCDNKey(match[1])] = match[2]
	}
	return entries, nil
}
